lib/Sema/CSSimplify.cpp

```diff
@@ -10864,6 +10864,12 @@ void ConstraintSystem::recordPotentialHole(Type type) {
   });
 }
 
+void ConstraintSystem::recordMatchCallArgumentResult(
+    ConstraintLocator *locator, MatchCallArgumentResult result) {
+  assert(locator->isLastElement<LocatorPathElt::ApplyArgument>());
+  argumentMatchingChoices.push_back({locator, result});
```

Comment:
I wonder whether it possible to assert that we don't record multiple call-matching results for the same locator?

Reply:
I think we could do that in `#ifndef NDEBUG` block.

Reply:
@hamishknight I actually did add that assertion initially and it hit it, so we are recording the same locator multiple times, but it's always with the same value. There's an existing assertion (modified in this pr for the change from TrailingClosureMatching to MatchCallArgumentResult) when adding them to the solution in `ConstraintSystem::finalize()` that allows duplicates if they have the same value as well:
```
// Remember all of the argument/parameter matching choices we made.
  for (auto &argumentMatch: argumentMatchingChoices) {
    auto inserted = solution.argumentMatchingChoices.insert(argumentMatch);
    assert(inserted.second || inserted.first->second == argumentMatch.second);
    (void)inserted;
  }
```
A comment just above that (for copying the DisjunctionChoices) has this comment:
```
  // Remember all the disjunction choices we made.
  for (auto &choice : DisjunctionChoices) {
    // We shouldn't ever register disjunction choices multiple times,
    // but saving and re-applying solutions can cause us to get
    // multiple entries.  We should use an optimized PartialSolution
    // structure for that use case, which would optimize a lot of
    // stuff here.
    assert(!solution.DisjunctionChoices.count(choice.first) ||
           solution.DisjunctionChoices[choice.first] == choice.second);
    solution.DisjunctionChoices.insert(choice);
  }
```
I assumed the same applied to this and didn't look into it any further. Did you want me to move (or copy) the check from finalize into this method?

Reply:
This makes sense to me, as soon as it's a re-addition of the same information it shouldn't be a problem.

Reply:
Ah okay, I see! I think the assertion in `finalize` is probably sufficient then, but it would be nice in the future to tighten up the assertions a bit when we stop applying solutions back to the constraint system before CSApply.